### ISSUE TITLE
Deb822 source support multi signature?

### ISSUE THREAD

**Issue by eatradish** [jelmer/deb822-rs#203]
I noticed that line style sources.list seems to support specifying multiple `signed-by` paths:

```
deb [ arch=amd64 signed-by=/usr/share/keyrings/termius-2023.gpg,/usr/share/keyrings/termius-2026.gpg ] https://deb.termius.com squeeze main
```

Then I noticed that `apt_sources` defines signature as `Option<Signature>` instead of `Option<Vec<Signature>>`

**Comment by jelmer** [comment]
At the moment apt-sources only supports rfc822 style apt sources files, and not traditional style files. We should support the latter too (including with multiple signature files).

**Comment by michalfita** [comment]
@eatradish the quoted example is `.list` file line, not `.sources`, and it's not DEB822 compatible. I have my private parser for them, but it's not in state to be released and IMVHO shall not be part of this repo.

The bug is on me. I've double check `sources.list` man page and it state there can be multiple files specified. It's going to be hell of parsing to recognize between _files_, _PGP blocks_ and _fingerprints_.

**Comment by jelmer** [comment]
FWIW I do want to support the older .sources.list files as well.

**Comment by eatradish** [comment]
Sorry, there seems to be some misunderstanding here ...... What I meant was that I saw that sources.list can support specifying multiple signature files and wanted to ask if the `.source` format is also supported, like this:


```
Types: deb
URIs: https://mirrors.bfsu.edu.cn/ubuntu
Suites: noble noble-updates noble-backports
Components: main restricted universe multiverse
Signed-By: /usr/share/keyrings/ubuntu-archive-keyring.gpg /path/to/key2.gpg   
```

**Comment by eatradish** [comment]
> Sorry, there seems to be some misunderstanding here ...... What I meant was that I saw that `sources.list` can support specifying multiple signature files and wanted to ask if the `.source` format is also supported, like this:
> 
> ```
> Types: deb
> URIs: https://mirrors.bfsu.edu.cn/ubuntu
> Suites: noble noble-updates noble-backports
> Components: main restricted universe multiverse
> Signed-By: /usr/share/keyrings/ubuntu-archive-keyring.gpg /path/to/key2.gpg   
> ```

I asked another person who is familiar with Ubuntu, and he told me that the Deb822 source supports specifying multiple signature files, and I wrote support for that in `oma-apt-sources-list`, which supports parse of both sources.list and deb822 files.

So when it comes to this part, I copied the `apt-sources` source and changed it to this:

```rust
/// A type to store
#[derive(Debug, PartialEq, Eq, Clone, Hash)]
pub enum Signature {
    /// The PGP key is stored inside the `.sources` files
    KeyBlock(String), // TODO: shall we validate PGP Public Key?
    /// The public key is store in a file of the given path
    KeyPath(Vec<PathBuf>), // TODO: man page specifies fingerprints, but there's no example
}
```

https://github.com/AOSC-Dev/oma-apt-sources-lists/blob/6149b3e1b074e5d8af5805c3c47e9666fd34d43c/src/deb822/signature.rs#L9

**Comment by michalfita** [comment]
> Sorry, there seems to be some misunderstanding here

OK, you put that as an example.

> I asked another person who is familiar with Ubuntu, and he told me that the Deb822 source supports specifying multiple signature files

I wrote in my comment that it is a bug and I confirmed with man pages that it should be supported. But this core is in early alpha stage not production usable yet. It doesn't for example support fingerprints mentioned in the man pages as I can't find unit tests for them in the original `apt` / `apt-get` code.

Why don't you PR sensible fix with unit tests?

The `apt-sources` subcrate of this project aims at exactly the same as PoPOS's `apt-sources-list` crate aim at in terms of delivering abstraction - I don't see reason why double the effort for the same ultimate goal?

**Comment by eatradish** [comment]
> > Sorry, there seems to be some misunderstanding here
> 
> OK, you put that as an example.
> 
> > I asked another person who is familiar with Ubuntu, and he told me that the Deb822 source supports specifying multiple signature files
> 
> I wrote in my comment that it is a bug and I confirmed with man pages that it should be supported. But this core is in early alpha stage not production usable yet. It doesn't for example support fingerprints mentioned in the man pages as I can't find unit tests for them in the original `apt` / `apt-get` code.
> 
> Why don't you PR sensible fix with unit tests?
> 
> The `apt-sources` subcrate of this project aims at exactly the same as PoPOS's `apt-sources-list` crate aim at in terms of delivering abstraction - I don't see reason why double the effort for the same ultimate goal?

Ah the reason for copying the source code to `oma-apt-sources-lists` is simply because the `apt-sources` crate was not uploaded to crates.io at the time.

**Comment by michalfita** [comment]
Right.

You're aware that cargo can use crate directly from the repository? [[Specifying dependencies from git repositories](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-git-repositories)](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-git-repositories)

**Comment by jelmer** [comment]
> > I do want to support the older .sources.list files as well.
> 
> Despite them **NOT** being DEB822? I have some time over Christmas, I may try to extract my _stuff_ into a branch, but it's ugly and read only (effectively my program took any `.list` files and if there were any changes to `Repository` replaced them with `.sources` file.

Yes - I think they're simple enough that it makes sense to support them. A lot of users will want both. Maybe a little bit odd for it to be living in a repository called "deb822-rs" but that shouldn't be a practical issues - worst case we can split it out into its own repo.

First step is perhaps to support both with different structs, then see if we can provide a common interface for them somehow (like you say, either the same struct, an enum or a trait?)

**Comment by michalfita** [comment]
@jelmer mine worked deserialization on the same, that format simply doesn't support most of optional fields in `.sources` format.

**Comment by jelmer** [comment]
Which fields are not supported? Per the source.list manpage, most fields have an equivalent in both file formats.

**Comment by michalfita** [comment]
@jelmer I'm not sure now which exactly aren't supported, but two major difference with one line `.list` format and DEB822 `.sources` format are:

1. One line can be either `deb` or `deb-src`, never both; `Types` field of DEB822 allows both to be present at the same time.
2. One line accepts one suite/distribution only; `Suites:` field of DEB822 format allows more (if that ever makes sense)
